## Assignment 3

##### Adishree Sane, 702207239, adsane@syr.edu

### Experiment 1

In [8]:
import pandas as pd
import numpy as np
import re
import nltk
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from nltk import pos_tag
from nltk.stem import WordNetLemmatizer
from sklearn.feature_extraction.text import TfidfVectorizer
from nltk.stem import PorterStemmer, WordNetLemmatizer
from nltk.tokenize import word_tokenize
from sklearn.feature_extraction.text import CountVectorizer

In [2]:

nltk.download('stopwords')
nltk.download('punkt')
nltk.download('wordnet')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\adiun\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\adiun\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\adiun\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [3]:

data = pd.read_csv('twitter-sanders-apple2.csv')
print(data.head())

  class                                               text
0   Pos  Now all @Apple has to do is get swype on the i...
1   Pos  @Apple will be adding more carrier support to ...
2   Pos  Hilarious @youtube video - guy does a duet wit...
3   Pos  @RIM you made it too easy for me to switch to ...
4   Pos  I just realized that the reason I got into twi...


In [4]:

stemmer = PorterStemmer()
lemmatizer = WordNetLemmatizer()


def preprocess_text(text):
    
    text = text.lower()

    text = re.sub(r'http\S+|www\S+', '', text)
    
  
    text = re.sub(r'<.*?>', '', text)
    
  
    text = re.sub(r'[^\x00-\x7F]+', '', text)
    

    text = re.sub(r'[^\w\s]', '', text)
    

    words = word_tokenize(text)
    
    
    stop_words = set(stopwords.words('english'))
    words = [word for word in words if word not in stop_words]

    words = [word for word in words if len(word) > 2 and len(word) < 20]
    

    words = [stemmer.stem(word) for word in words]
    

    text = ' '.join(words)
    
    return text

In [5]:
import nltk
nltk.download('punkt_tab')

[nltk_data] Downloading package punkt_tab to
[nltk_data]     C:\Users\adiun\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!


True

In [6]:
data['cleaned_text'] = data['text'].apply(preprocess_text)

print(data[['text', 'cleaned_text']].head())

                                                text  \
0  Now all @Apple has to do is get swype on the i...   
1  @Apple will be adding more carrier support to ...   
2  Hilarious @youtube video - guy does a duet wit...   
3  @RIM you made it too easy for me to switch to ...   
4  I just realized that the reason I got into twi...   

                                        cleaned_text  
0                   appl get swype iphon crack iphon  
1              appl ad carrier support iphon announc  
2  hilari youtub video guy duet appl siri pretti ...  
3                rim made easi switch appl iphon see  
4          realiz reason got twitter ios5 thank appl  


In [9]:

vectorizer_unigrams = CountVectorizer(binary=True, min_df=3, stop_words='english')
X = vectorizer_unigrams.fit_transform(data['cleaned_text'])
y = data['class'].apply(lambda x: 1 if x == 'Pos' else 0)  # Convert 'Pos' and 'Neg' to 1 and 0

In [10]:

print("Feature Matrix Shape:", X.shape)

print("Top 10 Unigram Features:", vectorizer_unigrams.get_feature_names_out()[:10])


Feature Matrix Shape: (479, 312)
Top 10 Unigram Features: ['3g' 'absolut' 'actual' 'air' 'album' 'alreadi' 'alway' 'amaz' 'android'
 'annoy']


In [11]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.svm import SVC
from sklearn.metrics import classification_report
from sklearn.feature_extraction.text import TfidfVectorizer

In [12]:
print(data.columns)


Index(['class', 'text', 'cleaned_text'], dtype='object')


In [13]:

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


nb_model = MultinomialNB()
nb_model.fit(X_train, y_train)


svm_model = SVC(kernel='linear')
svm_model.fit(X_train, y_train)


y_pred_nb = nb_model.predict(X_test)
print("Naïve Bayes Classification Report:")
print(classification_report(y_test, y_pred_nb))


y_pred_svm = svm_model.predict(X_test)
print("SVM Classification Report:")
print(classification_report(y_test, y_pred_svm))

Naïve Bayes Classification Report:
              precision    recall  f1-score   support

           0       0.88      0.80      0.84        61
           1       0.70      0.80      0.75        35

    accuracy                           0.80        96
   macro avg       0.79      0.80      0.79        96
weighted avg       0.81      0.80      0.80        96

SVM Classification Report:
              precision    recall  f1-score   support

           0       0.79      0.82      0.81        61
           1       0.67      0.63      0.65        35

    accuracy                           0.75        96
   macro avg       0.73      0.72      0.73        96
weighted avg       0.75      0.75      0.75        96



### Exeriment 2

In [18]:
!pip install spacy

  Using cached annotated_types-0.7.0-py3-none-any.whl.metadata (15 kB)
   ---------------------------------------- 0.0/12.2 MB ? eta -:--:--
   ------------------ --------------------- 5.8/12.2 MB 35.3 MB/s eta 0:00:01
   ---------------------------------------  12.1/12.2 MB 36.0 MB/s eta 0:00:01
   ---------------------------------------- 12.2/12.2 MB 29.4 MB/s eta 0:00:00
   ---------------------------------------- 0.0/1.9 MB ? eta -:--:--
   ---------------------------------------- 1.9/1.9 MB 54.4 MB/s eta 0:00:00
   ---------------------------------------- 0.0/632.6 kB ? eta -:--:--
   --------------------------------------- 632.6/632.6 kB 12.0 MB/s eta 0:00:00
   ---------------------------------------- 0.0/1.5 MB ? eta -:--:--
   ---------------------------------------- 1.5/1.5 MB 26.1 MB/s eta 0:00:00
Using cached annotated_types-0.7.0-py3-none-any.whl (13 kB)
   ---------------------------------------- 0.0/6.2 MB ? eta -:--:--
   ---------------------------------------- 6.2/6.2

In [19]:
!python -m spacy download en_core_web_sm

     ---------------------------------------- 0.0/12.8 MB ? eta -:--:--
     ------ --------------------------------- 2.1/12.8 MB 19.7 MB/s eta 0:00:01
     ------------------------------- ------- 10.2/12.8 MB 33.5 MB/s eta 0:00:01
     --------------------------------------- 12.8/12.8 MB 32.2 MB/s eta 0:00:00
[+] Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')


In [20]:
import spacy

In [21]:

def pos_tagging_spacy(text):
    doc = nlp(text)
    # Create POS tag representation "word_POS"
    return " ".join([f"{token.text}_{token.pos_}" for token in doc])

In [23]:

nlp = spacy.load("en_core_web_sm")


In [24]:

data['pos_text_spacy'] = data['cleaned_text'].apply(pos_tagging_spacy)
print(data[['text', 'pos_text_spacy']].head())

                                                text  \
0  Now all @Apple has to do is get swype on the i...   
1  @Apple will be adding more carrier support to ...   
2  Hilarious @youtube video - guy does a duet wit...   
3  @RIM you made it too easy for me to switch to ...   
4  I just realized that the reason I got into twi...   

                                      pos_text_spacy  
0  appl_ADV get_VERB swype_NOUN iphon_NOUN crack_...  
1  appl_VERB ad_NOUN carrier_NOUN support_PROPN i...  
2  hilari_PROPN youtub_PROPN video_NOUN guy_NOUN ...  
3  rim_NOUN made_VERB easi_ADJ switch_NOUN appl_P...  
4  realiz_NOUN reason_NOUN got_VERB twitter_ADJ i...  


In [32]:
vectorizer_pos = CountVectorizer(binary=True, stop_words='english')
X_pos = vectorizer_pos.fit_transform(data['pos_text_spacy'])


In [33]:

print("POS Feature Matrix Shape:", X_pos.shape)

print("Top 10 POS Features:", vectorizer_pos.get_feature_names_out()[:10])


POS Feature Matrix Shape: (479, 1878)
Top 10 POS Features: ['100_num' '135_num' '199_num' '1hr_propn' '1st_adj' '1st_propn'
 '1stgen_num' '2001_num' '200_num' '2012_num']


In [34]:
X_train_pos, X_test_pos, y_train_pos, y_test_pos = train_test_split(X_pos, y, test_size=0.2, random_state=42)

nb_model_pos = MultinomialNB()
nb_model_pos.fit(X_train_pos, y_train_pos)

svm_model_pos = SVC(kernel='linear')
svm_model_pos.fit(X_train_pos, y_train_pos)

y_pred_nb_pos = nb_model_pos.predict(X_test_pos)
print("Naïve Bayes Classification Report (POS):")
print(classification_report(y_test_pos, y_pred_nb_pos))

y_pred_svm_pos = svm_model_pos.predict(X_test_pos)
print("SVM Classification Report (POS):")
print(classification_report(y_test_pos, y_pred_svm_pos))

Naïve Bayes Classification Report (POS):
              precision    recall  f1-score   support

           0       0.91      0.79      0.84        61
           1       0.70      0.86      0.77        35

    accuracy                           0.81        96
   macro avg       0.80      0.82      0.81        96
weighted avg       0.83      0.81      0.82        96

SVM Classification Report (POS):
              precision    recall  f1-score   support

           0       0.82      0.84      0.83        61
           1       0.71      0.69      0.70        35

    accuracy                           0.78        96
   macro avg       0.76      0.76      0.76        96
weighted avg       0.78      0.78      0.78        96



### Results

#### Naïve Bayes (Unigrams vs POS features):
##### Unigrams:

Negative Class: Precision = 0.88, Recall = 0.80, F1-Score = 0.84

Positive Class: Precision = 0.70, Recall = 0.80, F1-Score = 0.75

Accuracy: 0.80

##### POS Features:

Negative Class: Precision = 0.91, Recall = 0.79, F1-Score = 0.84

Positive Class: Precision = 0.70, Recall = 0.86, F1-Score = 0.77

Accuracy: 0.81

##### Conclusion:

POS features improve Naïve Bayes performance for negative and positive sentiment, with an increase in recall for positive sentiment.



#### SVM (Unigrams vs POS features):
##### Unigrams:

Negative Class: Precision = 0.79, Recall = 0.82, F1-Score = 0.81

Positive Class: Precision = 0.67, Recall = 0.63, F1-Score = 0.65

Accuracy: 0.75

##### POS Features:

Negative Class: Precision = 0.82, Recall = 0.84, F1-Score = 0.83

Positive Class: Precision = 0.71, Recall = 0.69, F1-Score = 0.70

Accuracy: 0.78

##### Conclusion:

POS features improve SVM performance for negative sentiment, but the increase for positive sentiment is average.



#### Overall Observations:
Naïve Bayes benefits more from POS features than SVM regarding accuracy and recall for positive sentiment.

SVM shows improvement with POS features in negative sentiment and overall accuracy.

POS features improve precision and recall for negative sentiment in both models.